iModelCore/GeomLibs/serialization/src/FlatBuffer/FixedStructs.cpp

```diff
@@ -1918,89 +1922,92 @@ static void ReadVariantGeometry(const BGFB::VariantGeometry * fbGeometry, bvecto
             dest.push_back(g);
         }
     }
-};    
-    
-IGeometryPtr BentleyGeometryFlatBuffer::BytesToGeometry (bvector <Byte> &buffer, bool applyValidation)
+};
+
+bool BentleyGeometryFlatBuffer::IsFlatBufferFormat(bvector<Byte> & buffer)
     {
-    if (buffer.size () == 0)
+        if (buffer.size() == 0)
+            return false;
+        Byte *fbStart = GetFBStart(buffer);
+        return nullptr != fbStart;
+    }
+
+bool BentleyGeometryFlatBuffer::IsFlatBufferFormat(Byte const *buffer)
+    {
+        if (nullptr == buffer)
+            return false;
+        Byte const *fbStart = GetFBStart(buffer);
+        return nullptr != fbStart;
+    }
+
+template <typename PtrType, typename ReaderMethodType>
+PtrType BytesToXXX(Byte const *buffer, size_t const bufferSize, bool applyValidation, ReaderMethodType readerMethod)
+    {
+    if (nullptr == buffer)
         return nullptr;
-    Byte* fbStart = GetFBStart (buffer);
+    Byte const *fbStart = GetFBStart(buffer);
     if (nullptr == fbStart)
-        return nullptr;        
-    auto fbRoot = flatbuffers::GetRoot <BGFB::VariantGeometry>(fbStart);
-    if (nullptr != fbRoot
-        && fbRoot->has_geometry_type ()
-        && fbRoot->has_geometry ()
-        )
+        return nullptr;
+    auto fbRoot = flatbuffers::GetRoot<BGFB::VariantGeometry>(fbStart);
+    auto myVerifier = flatbuffers::Verifier(fbStart, bufferSize - s_prefixBufferSize);
+    if (!fbRoot->Verify(myVerifier))
+        return nullptr;
+    if (nullptr != fbRoot && fbRoot->has_geometry_type() && fbRoot->has_geometry())
         {
-        auto result = FBReader::ReadGeometry (fbRoot);
-        if (applyValidation && s_readValidator.IsValid())
+        PtrType result = readerMethod(fbRoot);
+        if (result.IsValid())
             {
-            if (!result->IsValidGeometry(s_readValidator))
-                return nullptr;
+            if (applyValidation && s_readValidator.IsValid())
+                if (!result->IsValidGeometry(s_readValidator))
+                    return nullptr;

```

Comment:
Style tip: if an `if` statement is followed by a block of more than one line, use curly braces:

```suggestion
            if (applyValidation && s_readValidator.IsValid())
                {
                if (!result->IsValidGeometry(s_readValidator))
                    return nullptr;
                }
```